### DISCUSSION TITLE
CUDA Error

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2258] [Q&A] [answered]
Hi,
 
I tried to run classification on my own dataset on colab, unfortunately I have a problem with the GPU, I get this error:
"RuntimeError: CUDA error: device-side assert triggered"
I also deleted pin_memory=torch.cuda.is_available() from the loaders

Here is the code:
https://github.com/Meddebma/pyradiomics/blob/master/Projekt2.ipynb

Could you help me with this? 
Thank you

**Comment by rijobro** [comment, marked as answer]
Hi @Meddebma, I moved your tutorials issue to a discussion on the main repository. I think CUDA side-assert errors can occur when, for example, indexing out of range. In this case, I think debugging can be done by working on the CPU. I know this will be slow, but hopefully you'll have to run it for the first epoch. You can make your images considerably smaller (even 2d) if that would help.

**Reply by Meddebma** [reply]
Hi @rijobro, 
thanks for the quick answer, 
I also tried it with CPU, but I've got this error during the 1st epoch: "IndexError: Target 2 is out of bounds."
the images are abdominal CT-Scans with circa 100 slices, do you think I can choose a Region of Interest? for example looking only in the left hemiabdomen? or how can I make my images smaller?
Thanks

**Reply by rijobro** [reply]
Ok, so as expected it's an indexing error. What values do you have in your ground truth segmentation (`torch.unique`)? I notice you use the default resampling of `resize`, which is `area`. I'm not really sure how that works, but normally for labels you want to use `nearest` neighbour.

**Reply by Meddebma** [reply]
@rijobro I did not use any segmentation yet, I just converted DICOMs of abdominal CT-Scans to NIFTI (nii.gz) of patients with two differents diseases and wanted to see if classification could work there. 

when I put torch.unique(im) I get:
tensor([0.0000, 0.4219, 0.4221,  ..., 0.9724, 0.9771, 0.9784])

**Reply by rijobro** [reply]
Oh sorry, I missed that. Looks like your labels are 1s and 2s. Try switching them to 0s and 1s.

**Reply by Meddebma** [reply]
@rijobro oh crazy!! it worked! thank you!

**Reply by Meddebma** [reply]
@rijobro 
I would like to visualize the dataset after transforms, I tried to plot the dataloader but it did not work, do you know how is the correct code to do that? Also when I started training it seemed that something with some CT_scans is wrong, but I cant locate the wrong file, how can I find out which Nifti Files contain errors? 
thank you so much

**Reply by rijobro** [reply]
What do you mean you tried to plot the dataloader? What errors are you seeing for the nifti files?

**Reply by Meddebma** [reply]
@rijobro 
By plotting I meant showing from every nifty file one slice after transforms, as it is done under "Check data shape and visualize" in this tutorial: https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/brats_segmentation_3d.ipynb

the error in training is to find in under training section:
https://github.com/Meddebma/pyradiomics/blob/master/Projekt2.ipynb

thanks

**Reply by rijobro** [reply]
Looks like one of your images already has a channel dimension: `Shape: (1, 512, 512, 54, 2)`. So for this one, instead of `AddChannel`, you should use `AsChannelFirst`. If some of your images have a channel dimension already, and some don't then you'll have to use if/else logic.

**Reply by Meddebma** [reply]
@rijobro 
I wrote something like this but it didnt work 

check_data = first(check_loader)
image= check_data["image"][0][0]
train_transforms_1 = Compose([ScaleIntensity(), AddChannel(), Resize((96, 96, 96),mode="nearest"), ToTensor()])
train_transforms_2 = Compose([ScaleIntensity(), AsChannelFirst(), Resize((96, 96, 96),mode="nearest"), ToTensor()])
for i in image:
    if len(image.shape)>4:
        train_ds = ImageDataset(image_files=images[:100], labels=labels[:100], transform=train_transforms_2)
    else:
        train_ds = ImageDataset(image_files=images[:100], labels=labels[:100], transform=train_transforms_1) 
it is a bit complex for me sorry :(

**Reply by rijobro** [reply]
the problem is that `check_data["image"][0]` will only check the first image, and then assumes all the others after are the same. Which appears not to be the case.

There may be a better way to do this, but my original thought would be to create a custom transform that either adds a channel or moves the channel, depending on the dimensionality of the image. Something like (not tested, just to give you an idea):

```python
class ChannelFirstOrAdd(Transform):
    def __init__(self):
        self.channel_first = AsChannelFirst()
        self.add_channel = AddChannel()

    def __call__(self, im):
        ndims = im.ndim
        if ndims == 4:
            return self.channel_first(im)
        elif ndims == 3:
            return self.add_channel(im)
        else:
            raise RuntimeError

train_transforms = Compose([ScaleIntensity(), ChannelFirstOrAdd(), Resize((96, 96, 96),mode="nearest"), ToTensor()])
```

**Reply by Meddebma** [reply]
@rijobro 
Ah that's a very good Idea! 
 I added 

from monai.transforms import Transform 

but then I got this Error:

TypeError: Can't instantiate abstract class ChannelFirstOrAdd with abstract methods __call__

**Reply by rijobro** [reply]
did you use `def __call__` or `def call`? Note that I updated the answer to use `__call__`. For me the following works:

```python
import numpy as np
from monai.transforms import Transform, AddChannel, AsChannelFirst
class ChannelFirstOrAdd(Transform):
    def __init__(self):
        self.channel_first = AsChannelFirst()
        self.add_channel = AddChannel()

    def __call__(self, im):
        ndims = im.ndim
        if ndims == 4:
            return self.channel_first(im)
        elif ndims == 3:
            return self.add_channel(im)
        else:
            raise RuntimeError

t = ChannelFirstOrAdd()
already_has_channel = np.zeros((3, 4, 5, 2))
print(t(already_has_channel).shape)

needs_channel_added = np.zeros((3, 4, 5))
print(t(needs_channel_added).shape)
```
Output is:
```
(2, 3, 4, 5)
(1, 3, 4, 5)
```
That is to say that in the first case, the 2 was moved to the front, and in the second case, a 1 was added to the front.

**Reply by Meddebma** [reply]
@rijobro 
Yes now it worked! 
Now the Training is going forward, but I got this new Error:
RuntimeError: stack expects each tensor to be equal size, but got [1, 96, 96, 96] at entry 0 and [2, 96, 96, 96] at entry 1
does this have again to do with the first channel?

**Reply by Meddebma** [reply]
@rijobro 
here is the code with the new Error:
https://github.com/Meddebma/pyradiomics/blob/master/Projekt2.ipynb

**Reply by rijobro** [reply]
After the dataloader, a batch of images are concatenated together, which means they need to have dimensions that match. Some of your images have 1 channel, some have 2. This is what the error is trying to explain.

I guess you need to think about what the two dimensions of the CT image mean. You could either split them into two independent images, combine them together (e.g., mean) or discard one of the channels.

**Reply by Meddebma** [reply]
Ah now I get it! 
I think in some patients there were CT-Scans in different Kontrast-media phases, that is why there is an extra channel. How can I find out whether a channel is empty and discard one of the channels?

**Reply by Meddebma** [reply]
Dear @rijobro ,

I found out that some of the CT-Scans have this shape: torch.Size([1, 512, 512, 56, 1, 2],
I tried to remove the two last channels with squeezedim() but it did not work, here is what I tried:

```python
class Squeeze_ChannelFirst(Transform):
    def __init__(self):
        self.channel_first = AsChannelFirst()
        self.squeezedim = SqueezeDim()
    def __call__(self, im):
        ndims = im.ndim
        if ndims > 5:
            return self.squeezedim(dim=6)   
          elif ndims == 5:
            return self.squeezedim(dim=5)
        else:
            return self.channel_first(im)
```

is there another way to do it?

**Reply by rijobro** [reply]
Don't think you're using `SqueezeDim` correctly. The constructor should take `dim` as argument, and when you call it (`self.squeezedim`), you should pass `im` as argument. 

You could squeeze your image at the beginning to remove all excess dimensions eg: `[1, 320, 320, 320, 1, 2]` -> `[320, 320, 320, 2]`. And from there you use the same logic as before:

```python
class Squeeze_ChannelFirst(Transform):
    def __init__(self):
        self.channel_first = AsChannelFirst()
        self.add_channel = AddChannel()
        # squeeze all singleton dimensions eg: (1, 320, 320, 320, 1, 2) -> (320, 320, 320, 2)
        self.squeezedim = SqueezeDim(dim=None)
    def __call__(self, im):
        im = self.squeezedim(im)
        ndims = im.ndim
        if ndims == 4:
            return self.channel_first(im)
        elif ndims == 3:
            return self.add_channel(im)
        else:
            raise RuntimeError
```

**Reply by rijobro** [reply]
> How can I find out whether a channel is empty and discard one of the channels?

You'll have to create a transform that checks the two channels and return the one that you consider to be "non-empty". I can't tell you how to do that as I don't know what you mean by "empty". All zeros? NaNs?

**Reply by Meddebma** [reply]
@rijobro 
Thank you for your help, the problem is solved and all the data are now homogeneous :)
